lingk_scraper.py

```diff
@@ -0,0 +1,155 @@
+import re
+import datetime
+import requests
+import hmac
+import os
+import base64
+import urllib
+from hashlib import sha1
+import sys
+import json
+
+# ******************* These are copied from server.py ************************
+
+# We will separate these into a new helper file later when we separate the
+# server.py into modules.
+
+class ScrapeError(Exception):
+    pass
+
+## Logging
+
+def log(message):
+    print("[{}] {}".format(
+        datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+        message), file=sys.stderr)
+
+# ***************************************************************************
+
+## Constants
+
+BASE = 'www.lingkapis.com'
+EXTENSION = '/v1/harveymudd/coursecatalog/ps/datasets/coursecatalog'
+
+if 'KEY' in os.environ and 'SECRET' in os.environ:
+    KEY = os.environ['KEY']
+    SECRET = os.environ['SECRET']
+else:
+    raise Exception('KEY and SECRET need to be set in the environment.')
+
+## Helpers
+
+def time_html_format() -> str:
+    return datetime.datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S UTC')
+
+def gen_auth_header(key: str, secret: str, date_str: str) -> str:
+    '''
+    Returns an authentication header for HMAC-based authentication.
+    Credit to Jordan R.A., Olivia Watkins, Kate Woolverton, and Gavin Yancey
+    for figuring out the nuances of this awful, broken API.
+    '''
+    # whitespace-sensitive, requires the newline
+    sig = (f'date: {date_str}\n'
+           # method has to be lowercase
+           f'(request-target): get ') \
+           + EXTENSION
+    # convert to bytes for hmac
+    hashed_sig = hmac.new(bytes(secret, 'ascii'), bytes(sig, 'ascii'), sha1)
+    # convert to base64
+    encoded_sig = base64.b64encode(hashed_sig.digest())
+    # parse into url-friendly format
+    urlified_sig = urllib.parse.quote(encoded_sig)
+    return (f'Signature keyId="{key}",'
+            f'algorithm="hmac-sha1",'
+            'headers="date (request-target)",'
+            f'signature="{urlified_sig}"')
+
+## Course data scraping
+
+def get_course_data(key: str, secret: str, limit: int=10000,
+                    iterations: int=10) -> dict:
```

Comment:
Why do we do multiple iterations instead of having the server layer handle failures and retries like we do for Portal scraping (which also allows us to do exponential backoff if desired)?